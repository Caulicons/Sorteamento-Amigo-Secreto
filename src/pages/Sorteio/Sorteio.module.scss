@import 'styles/variaveis';

.form {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  row-gap: 25px;

  &_select {
    padding: 12px 16px;
    border-radius: 2rem;
    border: solid 2px #000;
    box-shadow: 4px 4px;
    width: 80%;
    max-width: 600px;
  }

  &_option {
    padding: 12px 16px;
    border-radius: 2rem;
    border: solid 2px #000;
    box-shadow: 4px 4px;
    width: 80%;
    max-width: 600px;
  }

  p {
    font-weight: 400;
    font-size: 16px;
    line-height: 27px;
    text-align: center;
  }

  &_button {
    width: 180px;
    padding: 16px;
    background-color: $laranja;
    color: $branco;
    font-weight: 600;
    font-size: 16px;
    border-radius: 2rem;
    border: solid 1px #000;
    box-shadow: 3px 3px 0px #000000;

    &:hover {
      cursor: pointer;
      background-color: $azul;
    }
  }

  &_resultado {
    font-weight: 400;
    font-size: 27px;
    line-height: 38px;
    color: $laranja;
  }
}

.aviaozinho {
  width: 120px;
  margin-top: 15px;
}

@media screen and (min-width: $width-desktop) {
  .form {

    &_select {
      padding: 24px;
    }

    &_button {
      padding: 24px
    }
  }

  .aviaozinho {
    width: auto;
    margin-top: 20px;
  }

}