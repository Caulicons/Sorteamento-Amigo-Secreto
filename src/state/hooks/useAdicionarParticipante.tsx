import { useRecoilValue, useSetRecoilState } from "recoil"
import { listaDeParticipantes, mensagemDeError } from "state/atoms"

export const useAdicionarParticipantes = () => {
  const setListaDeParticipantes = useSetRecoilState(listaDeParticipantes);
  const listaAtual = useRecoilValue(listaDeParticipantes);
  const setMensagemDeError = useSetRecoilState(mensagemDeError)

  return (nome: string) => {
    if(listaAtual.includes(nome)){
      setMensagemDeError('Nomes duplicados não são permitidos')
      setTimeout(()=> {
        setMensagemDeError('')
      }, 3500)
      return
    }
    return setListaDeParticipantes(listaAnterior => [...listaAnterior, nome])
  }
}