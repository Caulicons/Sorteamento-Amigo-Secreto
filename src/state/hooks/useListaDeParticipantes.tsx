import { useRecoilValue } from "recoil";
import { listaDeParticipantes } from "state/atoms";

const useListaDeParticipantes = () => {
  return useRecoilValue(listaDeParticipantes)
};

export default useListaDeParticipantes; 