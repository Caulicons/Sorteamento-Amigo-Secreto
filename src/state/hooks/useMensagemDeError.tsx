import { useRecoilValue } from "recoil"
import { mensagemDeError } from "state/atoms"

export const useMensagemDeError = () => {
  const mensagemError = useRecoilValue(mensagemDeError)
  return mensagemError
}