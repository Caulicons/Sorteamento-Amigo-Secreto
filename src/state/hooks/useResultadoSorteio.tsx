import { useRecoilValue } from "recoil"
import { resultadoSorteio } from "state/atoms"

export const useResultadoSorteio = () => {
  return useRecoilValue(resultadoSorteio)
}