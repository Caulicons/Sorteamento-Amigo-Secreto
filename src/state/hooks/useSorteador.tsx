import { realizarSorteio } from "helpers/resultadoSorteio/realizarSorteio";
import { useSetRecoilState } from "recoil";
import { resultadoSorteio } from "state/atoms";
import useListaDeParticipantes from "./useListaDeParticipantes";

export const useSorteador = () => {

  const participantes = useListaDeParticipantes()
  const setResposta = useSetRecoilState(resultadoSorteio)

  return () => {
    const resposta = realizarSorteio(participantes);
    setResposta(resposta)
  }
}
