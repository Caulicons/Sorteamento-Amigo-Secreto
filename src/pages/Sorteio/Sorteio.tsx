import { useState } from "react";
import useListaDeParticipantes from "state/hooks/useListaDeParticipantes"
import { useResultadoSorteio } from "state/hooks/useResultadoSorteio";
import styles from './Sorteio.module.scss';
import aviãoDePapel from 'assets/aviao.png'

const Sorteio = () => {

  const participantes = useListaDeParticipantes();
  const [amigoSecreto, setAmigoSecreto] = useState('');
  const [participante, setParticipante] = useState('')
  const sorteio = useResultadoSorteio()

  const mostrandoAmigoSecreto = (evento: React.FormEvent<HTMLFormElement>) => {
    evento.preventDefault();

    if (sorteio.has(participante)) {
      setAmigoSecreto(sorteio.get(participante)!)
    }
  }

  return (
    <>
      <form onSubmit={mostrandoAmigoSecreto} className={styles.form}>
        <select
          className={styles.form_select}
          name="selecionarParticipante"
          placeholder="Selecione seu nome"
          onChange={evento => setParticipante(evento.target.value)}
          value={participante}
        >
          <option>Selecione seu nome</option>
          {participantes.map(participante => <option  className={styles.form_option} value={participante} key={participante}>{participante}</option>)}
        </select>
        <button
          type='submit'
          className={styles.form_button}> Sortear!</button>
      </form>
      {amigoSecreto && <p
        className={styles.form_resultado}
        role='alert'
      >
        {amigoSecreto}</p>}

      <img src={aviãoDePapel} alt="Aviãozinho de papel" className={styles.aviaozinho}/>
    </ >
  )
}

export default Sorteio;