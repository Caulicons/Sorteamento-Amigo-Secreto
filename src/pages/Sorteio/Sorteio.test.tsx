import { fireEvent, render, screen } from "@testing-library/react";
import { RecoilRoot } from "recoil";
import useListaDeParticipantes from "state/hooks/useListaDeParticipantes";
import { useResultadoSorteio } from "state/hooks/useResultadoSorteio";
import Sorteio from './Sorteio';

jest.mock('state/hooks/useListaDeParticipantes')
jest.mock('state/hooks/useResultadoSorteio', () => {
  return {
    useResultadoSorteio: jest.fn()
  }
})

describe('na página de sorteio', () => {
  const participantes = [
    'amaral',
    'cabral',
    'lebral',
  ]

  const resultado = new Map([
    ['amaral', 'cabral'],
    ['cabral', 'jurimeu'],
    ['jurimeu', 'amaral']
  ])

  beforeEach(() => {
    (useListaDeParticipantes as jest.Mock).mockReturnValue(participantes);
    (useResultadoSorteio as jest.Mock).mockReturnValue(resultado)

  })

  test('todos participantes podem exibir seu amigo secreto.', () => {

    render(<RecoilRoot>
      <Sorteio />
    </RecoilRoot>)

    const opcoes = screen.queryAllByRole('option')
    expect(opcoes).toHaveLength(participantes.length + 1) /* O ' + 1' é por causa que já temos uma como padrão. */
  })

  test('o amigo secreto é exibido quando solicitado.', () => {
    render(<RecoilRoot>
      <Sorteio />
    </RecoilRoot>)

    const select = screen.getByPlaceholderText('Selecione seu nome')
    fireEvent.change(select, {
      target: {
        value: participantes[0]
      }
    })

    const button = screen.getByRole('button');
    fireEvent.click(button)

    const mensagem = screen.getByRole('alert');
    expect(mensagem).toBeInTheDocument()
  })

  test('O componente Sorteio deve conter essa estrutura', () => {

    const { container } = render(<RecoilRoot>
      <Sorteio />
    </RecoilRoot>)

    expect(container).toMatchSnapshot()
  })
})