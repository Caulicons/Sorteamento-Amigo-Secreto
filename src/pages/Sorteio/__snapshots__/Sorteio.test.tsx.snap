// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`na página de sorteio O componente Cadastro deve conter essa estrutura 1`] = `
<div>
  <form
    class="form"
  >
    <select
      class="form_select"
      name="selecionarParticipante"
      placeholder="Selecione seu nome"
    >
      <option>
        Selecione seu nome
      </option>
      <option
        class="form_option"
        value="amaral"
      >
        amaral
      </option>
      <option
        class="form_option"
        value="cabral"
      >
        cabral
      </option>
      <option
        class="form_option"
        value="lebral"
      >
        lebral
      </option>
    </select>
    <button
      class="form_button"
      type="submit"
    >
       Sortear!
    </button>
  </form>
  
  <img
    alt="Aviãozinho de papel"
    class="aviaozinho"
    src="aviao.png"
  />
</div>
`;

exports[`na página de sorteio O componente Sorteio deve conter essa estrutura 1`] = `
<div>
  <form
    class="form"
  >
    <select
      class="form_select"
      name="selecionarParticipante"
      placeholder="Selecione seu nome"
    >
      <option>
        Selecione seu nome
      </option>
      <option
        class="form_option"
        value="amaral"
      >
        amaral
      </option>
      <option
        class="form_option"
        value="cabral"
      >
        cabral
      </option>
      <option
        class="form_option"
        value="lebral"
      >
        lebral
      </option>
    </select>
    <button
      class="form_button"
      type="submit"
    >
       Sortear!
    </button>
  </form>
  
  <img
    alt="Aviãozinho de papel"
    class="aviaozinho"
    src="aviao.png"
  />
</div>
`;

exports[`na página de sorteio o código deve ficar dessa forma 1`] = `
<div>
  <form
    class="form"
  >
    <select
      class="form_select"
      name="selecionarParticipante"
      placeholder="Selecione seu nome"
    >
      <option>
        Selecione seu nome
      </option>
      <option
        class="form_option"
        value="amaral"
      >
        amaral
      </option>
      <option
        class="form_option"
        value="cabral"
      >
        cabral
      </option>
      <option
        class="form_option"
        value="lebral"
      >
        lebral
      </option>
    </select>
    <button
      class="form_button"
      type="submit"
    >
       Sortear!
    </button>
  </form>
  
  <img
    alt="Aviãozinho de papel"
    class="aviaozinho"
    src="aviao.png"
  />
</div>
`;
